### PR TITLE
Ensure Query can be updated

### PR DESCRIPTION
**Pull request by steipete**
When a query is replaced in a view, the `@StateObject` keeps the `Core` class alive, and it might still use the previous query. This additional bool tracks this case and ensures the base query is updated.

Example where this caused issues:
Screen Shot 2021-04-07 at 15 35 29

The query always ran with `userId = 0` (the early state while parent is resolving a suer) despite me updating the binding in init.

### CODE REVIEW

**Comment by groue** [comment]
🎉 Thank you @steipete!

I'll merge your PR and #956 is the next release :-)

**Comment by groue** [comment]
@steipete I'll try to reproduce your initial issue, so that I can increase my confidence that `@Query` does not break in the future.

May I ask you a question? Do you have any hint, a web page I could look at, so that I can start writing tests for those heavily SwiftUI-dependent objects?

**Comment by steipete** [comment]
Sure! I‘ll write up a PR today. Have a bit of experience in that area.

**Comment by steipete** [comment]
I've added two UI tests to the target. I'm not 100% happy about the "initialOrdering" and this could be a binding too, but it illustrates the issue - if we late-change the value after view has been displayed, the StateObject is created and (unless my fix) can't be changed anymore.

Hit me up on Twitter if you wanna chat, my DMs are open.

File: Documentation/DemoApps/GRDBCombineDemo/GRDBCombineDemo.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -696,6 +768,46 @@
...
+				TARGETED_DEVICE_FAMILY = "1,2";
+				TEST_TARGET_NAME = GRDBCombineDemo;
+			};
+			name = Debug;
+		};
+		78E2F9EB261EEC22005D1F7F /* Release */ = {
+			isa = XCBuildConfiguration;
+			buildSettings = {
+				CODE_SIGN_STYLE = Automatic;
+				DEVELOPMENT_TEAM = Y5PE65HELJ;
```
Comment (steipete) [inline Documentation/DemoApps/GRDBCombineDemo/GRDBCombineDemo.xcodeproj/project.pbxproj:795]: You'll want to change that

File: Documentation/DemoApps/GRDBCombineDemo/GRDBCombineDemo.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -696,6 +768,46 @@
...
+				INFOPLIST_FILE = GRDBCombineDemoUITests/Info.plist;
+				IPHONEOS_DEPLOYMENT_TARGET = 14.5;
+				LD_RUNPATH_SEARCH_PATHS = (
+					"$(inherited)",
+					"@executable_path/Frameworks",
+					"@loader_path/Frameworks",
+				);
+				PRODUCT_BUNDLE_IDENTIFIER = com.groue.GRDBCombineDemoUITests;
```
Comment (steipete) [inline Documentation/DemoApps/GRDBCombineDemo/GRDBCombineDemo.xcodeproj/project.pbxproj:783]: I skipped the github part here, felt cleaner without.

File: Documentation/DemoApps/GRDBCombineDemo/GRDBCombineDemoUITests/GRDBCombineDemoUITests.swift [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+        // Given default sorting, Henriette must be on top
+        XCTAssertEqual(getFirstCell().label, "Craig, 8 points")
+
+        // Reverse sorting will bring Arthur on top (sort by name)
+        app.buttons["Score"].tap()
+        XCTAssertEqual(getFirstCell().label, "Arthur, 5 points")
+    }
+
+    func testListAndAddPlayer() throws {
```
Comment (steipete) [inline Documentation/DemoApps/GRDBCombineDemo/GRDBCombineDemoUITests/GRDBCombineDemoUITests.swift:44]: This test is unrelated and just tests basic add functionality.

File: Documentation/DemoApps/GRDBCombineDemo/GRDBCombineDemoUITests/GRDBCombineDemoUITests.swift [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+    func testInitialSortingIsByScore() throws {
+        let app = runApp()
+
+        let getFirstCell = { () -> XCUIElement in
+            app.tables.element(boundBy: 0).cells.element(boundBy: 0)
+        }
+
+        // Given default sorting, Henriette must be on top
+        XCTAssertEqual(getFirstCell().label, "Craig, 8 points")
```
Comment (steipete) [inline Documentation/DemoApps/GRDBCombineDemo/GRDBCombineDemoUITests/GRDBCombineDemoUITests.swift:37]: This test fails here without my query change.

File: Documentation/DemoApps/GRDBCombineDemo/GRDBCombineDemoUITests/GRDBCombineDemoUITests.swift (hunk 3 above)
Comment (groue) [reply Documentation/DemoApps/GRDBCombineDemo/GRDBCombineDemoUITests/GRDBCombineDemoUITests.swift:44]: It's quite educational, perfect for a demo app! 👍

File: Documentation/DemoApps/GRDBCombineDemo/GRDBCombineDemoUITests/GRDBCombineDemoUITests.swift (hunk 4 above)
Comment (groue) [reply Documentation/DemoApps/GRDBCombineDemo/GRDBCombineDemoUITests/GRDBCombineDemoUITests.swift:52]: Thank you so much, so much, for this. 

Now... I have ruined your efforts in https://github.com/groue/GRDB.swift/pull/955/commits/a5ffc52dbec8c6277ef1de70c0e1e7903f4c339a, in order to make the demo app... a demo app, and not a test case. This test is still there, but it no longer checks for your setup, because I have removed the conditions where your `@Query` bugfix expresses itself 😅 

But I hope I have clearly expressed what we are talking about in https://github.com/groue/GRDB.swift/pull/955/commits/c34615ced323aa5802a7b61afdb997b93f71e1a4. I quite intend to support your use case in the future, and setup stable tests.

Until then, this does not prevent this pull request from being merged.

**Comment by groue** [comment]
Welcome to the GRDB contributors, @steipete :-)

Would you accept an invitation to be able to push to the repository? This would not create any obligation, of course!

**Comment by steipete** [comment]
@groue I'd be honored! This is an important part of the app I'm building so I have a vested interest in helping making it better!

**Comment by groue** [comment]
Well, consider yourself onboarded :-)
